src/fluree/db/json_ld/transact.cljc

```diff
@@ -188,22 +192,33 @@
                          ;; means we cannot transact shacl in same txn as
                          ;;data and have it enforced.
                          (<? (get-subject-types db-before sid new-type-sids)))
-          class-shapes (<? (shacl/class-shapes db-before classes))
+
+          class-shapes   (<? (shacl/class-shapes db-before classes))
           referring-pids (when shacl-target-objects-of?
-                           (cond-> (map flake/p (<? (query-range/index-range db-before :opst = [sid])))
+                           (cond-> (<? (query-range/index-range db-before :opst = [sid] {:flake-xf (map flake/p)}))
                              referring-pid (conj referring-pid)))
-          pred-shapes (when (seq referring-pids)
-                        (<? (shacl/targetobject-shapes db-before referring-pids)))
-          shacl-map   (merge-with into class-shapes pred-shapes)
+          pred-shapes    (when (seq referring-pids)
+                           (<? (shacl/targetobject-shapes db-before referring-pids)))
+          shacl-shapes (into class-shapes pred-shapes)
+          [pid->shape->p-shapes
```

Comment:
I struggled a bit to understand what was going on in this `reduce`. I think maybe naming this `pid->shape->p-shapes` to something that conveys that these are advanced-validation shapes would help?

Reply:
Good point about it being an unwieldy read. What it's doing is assembling the data we need for advanced validation into a structure that enables fast retrieval when we need it.

I've pulled it into its own function and given it a docstring that explains what is going on. I do think I will keep the names of the result variables, though, since I think having the structure right there in the name helps understand how its used at the call site.